numpy/core/src/umath/ufunc_object.c

```diff
@@ -516,6 +517,55 @@ _apply_array_wrap(
 }
 
 
+/*
+ * TODO: Refactor this to be reused by ResultType, which includes the same.
+ *
+ * Checks if arrays (and "scalars"/O-d arrays) are in the same category.
+ * If the arrays are in the same or a larger category, we allow demotion
+ * of scalars to the lowest possible dtype.
+ * In that case, it returns the category defined above. The integer
+ * category is interesting. If we have a mixed loop, we may want
+ * to prefer the signed integer?
+ */
+static int
+should_use_min_scalar(int nop, PyArrayObject **op)
+{
+    int i, use_min_scalar, kind;
+    int all_scalars = 1, max_scalar_kind = -1, max_array_kind = -1;
+
+    /*
+     * Determine if there are any scalars, and if so, whether
+     * the maximum "kind" of the scalars surpasses the maximum
+     * "kind" of the arrays
+     */
+    use_min_scalar = 0;
+    if (nop > 1) {
+        for(i = 0; i < nop; ++i) {
+            kind = dtype_kind_to_simplified_ordering(
```

Comment:
nit: move the `int` declaration of `i` and `kind` to these two lines. We have C99 declaration rules now, and I see backporting this to 1.16 as very unlikely.